### PR TITLE
Added I18n Implementation & Examples

### LINKED ISSUE #31: I18n support
Provide a way to customize all the hardcoded strings in the element.

### PR DESCRIPTION
**Pull request by platosha**
Resolves #31

[Review on Reviewable](https://reviewable.io/reviews/vaadin/vaadin-upload/36)

### CODE REVIEW

**Comment by samiheikki** [comment]
Reviewed 1 of 4 files at r1.
Review status: 1 of 4 files reviewed at latest revision, 1 unresolved discussion, some commit checks broke.

---

*[vaadin-upload.html, line 385 \[r1\]](https://reviewable.io:443/reviews/vaadin/vaadin-upload/36#-KCJqmEamvzUTfVP07op:-KCJqmEbTn65hS4Awtzc:-395024366) ([raw file](https://github.com/vaadin/vaadin-upload/blob/b034403eb71e9a2a21a28fb4cd8821e67bcf40a1/vaadin-upload.html#L385)):*
missing semicolon

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/vaadin/vaadin-upload/36)*

**Comment by manolo** [comment]
In a quick view without diving so much in the code, my feeling is that this is much more than a MVP. 

I don't think we need here to consider pluralize, or even size units in TB, etc, format time, etc.

So I would remove all those _default functions which have a lot of code to maintain and leave the user to provide a couple of functions: format size, format time. Pluralize should be part of the code provided by the user, as well as units, etc.

---

Review status: 1 of 4 files reviewed at latest revision, 2 unresolved discussions, some commit checks broke.

---

*[demo/i18n.html, line 104 \[r1\]](https://reviewable.io:443/reviews/vaadin/vaadin-upload/36#-KCK4m7IfMFeTdY2pprH:-KCK4m7JhSRk0RNVZKQ3:-999065203) ([raw file](https://github.com/vaadin/vaadin-upload/blob/b034403eb71e9a2a21a28fb4cd8821e67bcf40a1/demo/i18n.html#L104)):*
It is shorter if you extract the number from the options:

return (number < 10 ? '0' : '') + number

Anyway this is an example

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/vaadin/vaadin-upload/36#-:-KCK9rYzRGBaYt8wwyGY:704257311)*

**Comment by manolo** [comment]
Review status: 1 of 4 files reviewed at latest revision, 6 unresolved discussions, some commit checks broke.

---

*[vaadin-upload.html, line 357 \[r1\]](https://reviewable.io:443/reviews/vaadin/vaadin-upload/36#-KCKF7P-HjiLyI93nMb3:-KCKF7P0XPw3Ej5WFuIe:154590806) ([raw file](https://github.com/vaadin/vaadin-upload/blob/b034403eb71e9a2a21a28fb4cd8821e67bcf40a1/vaadin-upload.html#L357)):*
Don't complicate the component with pluralize, it's so complex to give here a solution for users.

---

*[vaadin-upload.html, line 365 \[r1\]](https://reviewable.io:443/reviews/vaadin/vaadin-upload/36#-KCKEvpi3asy1bCDo1-Q:-KCKEvpjKu6JdgPY4C5n:-105241842) ([raw file](https://github.com/vaadin/vaadin-upload/blob/b034403eb71e9a2a21a28fb4cd8821e67bcf40a1/vaadin-upload.html#L365)):*
returning either bytes  or megabytes should be OK for the MVP

---

*[vaadin-upload.html, line 372 \[r1\]](https://reviewable.io:443/reviews/vaadin/vaadin-upload/36#-KCKEFyNY1rmxiOacCOm:-KCKEFyOuy1_5FB9IW0n:780055243) ([raw file](https://github.com/vaadin/vaadin-upload/blob/b034403eb71e9a2a21a28fb4cd8821e67bcf40a1/vaadin-upload.html#L372)):*
I see too much code to maintain as the default implementation.  I would just return the number of seconds, which will cover the MVP.  How days minutes seconds are formatted, in which order, what kind of unit, direction etc is something which should be done by a specialised i18n engine.

---

*[vaadin-upload.html, line 394 \[r1\]](https://reviewable.io:443/reviews/vaadin/vaadin-upload/36#-KCKCF5K9WuOeNBRRr3t:-KCKCF5K9WuOeNBRRr3u:2026906913) ([raw file](https://github.com/vaadin/vaadin-upload/blob/b034403eb71e9a2a21a28fb4cd8821e67bcf40a1/vaadin-upload.html#L394)):*
about & remaining should be in the i18n object.
What makes me wonder if we need some message formatter :  'about {1} remaining' 

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/vaadin/vaadin-upload/36)*

**Comment by manolo** [comment]
Review status: 1 of 4 files reviewed at latest revision, 7 unresolved discussions, some commit checks broke.

---

*[vaadin-upload.html, line 301 \[r1\]](https://reviewable.io:443/reviews/vaadin/vaadin-upload/36#-KCKKBSmhMfodW0dyqQy:-KCKKBSmhMfodW0dyqQz:-1524044250) ([raw file](https://github.com/vaadin/vaadin-upload/blob/b034403eb71e9a2a21a28fb4cd8821e67bcf40a1/vaadin-upload.html#L301)):*
We don't need to maintain units and conversions, customised  user functions should be enough

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/vaadin/vaadin-upload/36)*

**Comment by manolo** [comment]
Review status: 1 of 4 files reviewed at latest revision, 7 unresolved discussions, some commit checks broke.

---

*[vaadin-upload.html, line 394 \[r1\]](https://reviewable.io:443/reviews/vaadin/vaadin-upload/36#-KCKCF5K9WuOeNBRRr3t:-KCKRdXngYTjvthupAIh:833984084) ([raw file](https://github.com/vaadin/vaadin-upload/blob/b034403eb71e9a2a21a28fb4cd8821e67bcf40a1/vaadin-upload.html#L394)):*
I think to avoid dealing with message formatter, I would make this info simpler and leave the user to provide a function like formatETA(), being the default implementation 

 return i18n.remaining + ' ' + remaining

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/vaadin/vaadin-upload/36)*

**Comment by platosha** [comment]
Review status: 0 of 4 files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

*[vaadin-upload.html, line 301 \[r1\]](https://reviewable.io:443/reviews/vaadin/vaadin-upload/36#-KCKKBSmhMfodW0dyqQy:-KCLQj1H-dY6gcXMFNpn:607153330) ([raw file](https://github.com/vaadin/vaadin-upload/blob/b034403eb71e9a2a21a28fb4cd8821e67bcf40a1/vaadin-upload.html#L301)):*
Picking unit and converting to it is universal regardless of language. No need to require from translators to do the exactly same work. Leaving this code in the component.

---

*[vaadin-upload.html, line 357 \[r1\]](https://reviewable.io:443/reviews/vaadin/vaadin-upload/36#-KCKF7P-HjiLyI93nMb3:-KCLR4rnlvkqIxgLDEJI:-499146009) ([raw file](https://github.com/vaadin/vaadin-upload/blob/b034403eb71e9a2a21a28fb4cd8821e67bcf40a1/vaadin-upload.html#L357)):*
Done.

---

*[vaadin-upload.html, line 365 \[r1\]](https://reviewable.io:443/reviews/vaadin/vaadin-upload/36#-KCKEvpi3asy1bCDo1-Q:-KCLRZ4-nTm9Wf3A7CDB:-25454470) ([raw file](https://github.com/vaadin/vaadin-upload/blob/b034403eb71e9a2a21a28fb4cd8821e67bcf40a1/vaadin-upload.html#L365)):*
It's not hard to pick the unit and convert bytes, the amount of code is the same as with support for only B, KB and MB.

The rounding adds some minor overhead though, but gives a nicer result.

---

*[vaadin-upload.html, line 372 \[r1\]](https://reviewable.io:443/reviews/vaadin/vaadin-upload/36#-KCKEFyNY1rmxiOacCOm:-KCLQNsQWTPW2nRp0CkV:-400025287) ([raw file](https://github.com/vaadin/vaadin-upload/blob/b034403eb71e9a2a21a28fb4cd8821e67bcf40a1/vaadin-upload.html#L372)):*
Done. Default remaining time format was switched to [DD:]HH:MM:SS. Time units are no longer used, and not present in i18n property. 

---

*[vaadin-upload.html, line 385 \[r1\]](https://reviewable.io:443/reviews/vaadin/vaadin-upload/36#-KCJqmEamvzUTfVP07op:-KCLSQxIxMkJ6cFNlme4:-499146009) ([raw file](https://github.com/vaadin/vaadin-upload/blob/b034403eb71e9a2a21a28fb4cd8821e67bcf40a1/vaadin-upload.html#L385)):*
Done.

---

*[vaadin-upload.html, line 394 \[r1\]](https://reviewable.io:443/reviews/vaadin/vaadin-upload/36#-KCKCF5K9WuOeNBRRr3t:-KCLSTqGw_D1loekqcO0:-499146009) ([raw file](https://github.com/vaadin/vaadin-upload/blob/b034403eb71e9a2a21a28fb4cd8821e67bcf40a1/vaadin-upload.html#L394)):*
Done.

---

*[demo/i18n.html, line 104 \[r1\]](https://reviewable.io:443/reviews/vaadin/vaadin-upload/36#-KCK4m7IfMFeTdY2pprH:-KCLSbod7B5XU-4PEkD5:-439309662) ([raw file](https://github.com/vaadin/vaadin-upload/blob/b034403eb71e9a2a21a28fb4cd8821e67bcf40a1/demo/i18n.html#L104)):*
Done. The default and example formats were swapped, using this tip for the default format.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/vaadin/vaadin-upload/36)*

**Comment by manolo** [comment]
just one comment otherwise :lgtm: 

---

Review status: 0 of 4 files reviewed at latest revision, 3 unresolved discussions, some commit checks broke.

---

*[vaadin-upload.html, line 301 \[r1\]](https://reviewable.io:443/reviews/vaadin/vaadin-upload/36#-KCKKBSmhMfodW0dyqQy:-KCLywI_7_LKrwCL3H8-:1275242489) ([raw file](https://github.com/vaadin/vaadin-upload/blob/b034403eb71e9a2a21a28fb4cd8821e67bcf40a1/vaadin-upload.html#L301)):*
It's ok to me since the developer could replace the size array by their symbols via i18n object

---

*[vaadin-upload.html, line 365 \[r1\]](https://reviewable.io:443/reviews/vaadin/vaadin-upload/36#-KCKEvpi3asy1bCDo1-Q:-KCLaTPk30dKF9VVbvyF:831949484) ([raw file](https://github.com/vaadin/vaadin-upload/blob/b034403eb71e9a2a21a28fb4cd8821e67bcf40a1/vaadin-upload.html#L365)):*
The double tilde `~~` operator is widely used to save some code when rounding, it is the same than Math.floor but I think is enough

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/vaadin/vaadin-upload/36#-:-KCLzDCsr69NgJMGc0-C:-1251787138)*

**Comment by manolo** [comment]
Reviewed 2 of 4 files at r1, 1 of 2 files at r2.
Review status: 3 of 4 files reviewed at latest revision, 3 unresolved discussions, some commit checks broke.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/vaadin/vaadin-upload/36)*

**Comment by tomivirkki** [comment]
Review status: 3 of 5 files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[vaadin-upload.html, line 139 \[r3\]](https://reviewable.io:443/reviews/vaadin/vaadin-upload/36#-KCPpz1LURbkkApaZUcH:-KCPpz1Mgu10Y4xYwHeL:-1748979416) ([raw file](https://github.com/vaadin/vaadin-upload/blob/174e405ac87f22afb559fddf9b7a2e9e1560acbd/vaadin-upload.html#L139)):*
Since we use i18n property now as the "de-facto standard" for localizing our components, this should probably be removed; no need to have 2 different methods for changing the drop label

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/vaadin/vaadin-upload/36)*

**Comment by tomivirkki** [comment]
*[vaadin-upload.html, line 362 \[r3\]](https://reviewable.io:443/reviews/vaadin/vaadin-upload/36#-KCPtgqnMWFyqB7uqMwp:-KCPtgqnMWFyqB7uqMwq:-2049268523) ([raw file](https://github.com/vaadin/vaadin-upload/blob/174e405ac87f22afb559fddf9b7a2e9e1560acbd/vaadin-upload.html#L362)):*
Why not make formatSize and formatTime as functions in the i18n element and provide these functions as default values for them?

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/vaadin/vaadin-upload/36)*

**Comment by platosha** [comment]
Review status: 3 of 5 files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[vaadin-upload.html, line 362 \[r3\]](https://reviewable.io:443/reviews/vaadin/vaadin-upload/36#-KCPtgqnMWFyqB7uqMwp:-KCPuqPnfcaeMKebgDAD:-184480544) ([raw file](https://github.com/vaadin/vaadin-upload/blob/174e405ac87f22afb559fddf9b7a2e9e1560acbd/vaadin-upload.html#L362)):*
It's already supported. You can give `i18n.formatSize and i18n.formatTime` functions if you want to replace the default implementations. But unlike the other i18n subproperties, these callbacks are optional.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/vaadin/vaadin-upload/36)*

**Comment by manolo** [comment]
Reviewed 1 of 2 files at r2, 1 of 1 files at r4.
Review status: 4 of 5 files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[vaadin-upload.html, line 139 \[r3\]](https://reviewable.io:443/reviews/vaadin/vaadin-upload/36#-KCPpz1LURbkkApaZUcH:-KCPvXoNm6bHBgm_Qlm8:-1530358752) ([raw file](https://github.com/vaadin/vaadin-upload/blob/174e405ac87f22afb559fddf9b7a2e9e1560acbd/vaadin-upload.html#L139)):*
It's not a matter of changing the label, but customizing the area, for instance for setting an icon, etc

---

*[vaadin-upload.html, line 362 \[r3\]](https://reviewable.io:443/reviews/vaadin/vaadin-upload/36#-KCPtgqnMWFyqB7uqMwp:-KCPwCEUvACrgvEVNKrH:-338411405) ([raw file](https://github.com/vaadin/vaadin-upload/blob/174e405ac87f22afb559fddf9b7a2e9e1560acbd/vaadin-upload.html#L362)):*
 in date-picker we have the approach Tomi mention, I don't have big reasons for one or other option but to me is more difficult to read this way

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/vaadin/vaadin-upload/36)*

**Comment by tomivirkki** [comment]
*[vaadin-upload.html, line 362 \[r3\]](https://reviewable.io:443/reviews/vaadin/vaadin-upload/36#-KCPtgqnMWFyqB7uqMwp:-KCPx_0b7AKYf73CGbLG:1080508268) ([raw file](https://github.com/vaadin/vaadin-upload/blob/174e405ac87f22afb559fddf9b7a2e9e1560acbd/vaadin-upload.html#L362)):*
The current approach (with computed format functions and the _getCallbackFn) is quite verbose and hard to follow. If we want to follow the same approach as in Date Picker then these should be just default functions in the default i18n object.

If not, dropping "this._formatSize" and using something like:

file.totalStr = (this.i18n.formatSize || this._defaultFormatSize)(total);

would clear up the current approach.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/vaadin/vaadin-upload/36)*

**Comment by tomivirkki** [comment]
*[vaadin-upload.html, line 139 \[r3\]](https://reviewable.io:443/reviews/vaadin/vaadin-upload/36#-KCPpz1LURbkkApaZUcH:-KCPwpEgRdn4rllB0Lkz:-1557258195) ([raw file](https://github.com/vaadin/vaadin-upload/blob/174e405ac87f22afb559fddf9b7a2e9e1560acbd/vaadin-upload.html#L139)):*
Why is it important to be able to customize the content freely here but not in the #buttons sections for example?

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/vaadin/vaadin-upload/36)*

**Comment by manolo** [comment]
Reviewed 1 of 2 files at r3.
Review status: all files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[vaadin-upload.html, line 139 \[r3\]](https://reviewable.io:443/reviews/vaadin/vaadin-upload/36#-KCPpz1LURbkkApaZUcH:-KCQ-CpAXv61g3WiG1UW:515337391) ([raw file](https://github.com/vaadin/vaadin-upload/blob/174e405ac87f22afb559fddf9b7a2e9e1560acbd/vaadin-upload.html#L139)):*
We want to custumize any area even the file list but these two places are already done, so I don't see any reason for removing it and add it latter. It is already documented in this way in the docs PR

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/vaadin/vaadin-upload/36)*

**Comment by platosha** [comment]
Review status: 4 of 5 files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[vaadin-upload.html, line 362 \[r3\]](https://reviewable.io:443/reviews/vaadin/vaadin-upload/36#-KCPtgqnMWFyqB7uqMwp:-KCQ2EYmmGHyU6bbXKMd:384480473) ([raw file](https://github.com/vaadin/vaadin-upload/blob/174e405ac87f22afb559fddf9b7a2e9e1560acbd/vaadin-upload.html#L362)):*
Now done this way:

- `_formatSmth` methods are explicit methods, not computed function properties anymore
- `i18n.formatSmth` optional callbacks are discovered and called by `_formatSmth`
- removed `_getCallbackFn`

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/vaadin/vaadin-upload/36)*

**Comment by platosha** [comment]
Reviewed 1 of 1 files at r5.
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/vaadin/vaadin-upload/36)*

**Comment by platosha** [comment]
Reviewed 1 of 1 files at r7.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/vaadin/vaadin-upload/36)*
